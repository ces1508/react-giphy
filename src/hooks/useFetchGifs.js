import { useState, useEffect } from 'react'
import getGifs from '../helpers/getGifs'

const initialState = {
  data: [],
  loading: true,
  error: false
}

export const useFetchGifs = (category) => {
  const [state, setState] = useState(initialState)

  useEffect(() => {
    const makeRequest = async () => {
      getGifs(category)
      .then(data => {
        setState(state => ({
          ...state,
          data: [...data, ...state.data],
          error: false
        }))
      })
      .catch(e => {
        setState(state => ({
          ...state,
          error: true
        }))
      })
      .finally(() => {
        setState(state => ({
          ...state,
          loading: false
        }))
      })
    }
    makeRequest()
  }, [category])

  return state
}