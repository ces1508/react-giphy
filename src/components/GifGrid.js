import React from 'react'
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs'
import GifGridItem from './GifGridItem'
// import getGifs from '../helpers/getGifs'


const GifGrid = ({ category }) => {

  const { data: gifs, error, loading } = useFetchGifs(category)

  return (
    <div className="animate__animated animate__fadeInDownBig animate__delay-2s">
      <h3>{category}</h3>
      {  loading &&  <p>Cargando ...</p> }
      { error && <p>ocurrio un error cargando los gifs</p>}
      <div className="card-grid">
        {gifs.map( image => (
          <GifGridItem
            key={image.id}
            {...image}
          />
        ))}
      </div>
    </div>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}

export default GifGrid
