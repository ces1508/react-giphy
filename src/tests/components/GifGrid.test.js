import React from 'react'
import GifGrid from '../../components/GifGrid'
import { shallow } from 'enzyme'
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs')

describe('<GifGrid />', () => {

  const initialCategory = 'goku'
  useFetchGifs.mockReturnValue({
    data: [],
    loading: true,
    error: false
  })
  const wrapprer = shallow(<GifGrid category={initialCategory}/>)

  it('<GifGrid /> should be rendered', () => {
    expect(wrapprer).toMatchSnapshot()
  })

  it ('<GifGrid /> should be have a h3 with category as text', () => {
    const h3 = wrapprer.find('h3')

    expect(h3.text().trim()).toBe(initialCategory)
  })

  it ('<GifGrid /> should be have a loading text', () => {
    const loadingWrapper = wrapprer.find('p')

    expect(loadingWrapper.text()).toBe('Cargando ...')
  })

  it ('<GifGrid /> should have a p tag when useFetchGifs return an error', () => {

    useFetchGifs.mockReturnValue({
      data: [],
      loading: false,
      error: true
    })
    const wrapperWithError = shallow(<GifGrid category={initialCategory}/>)
    expect(wrapperWithError).toMatchSnapshot()

    const p = wrapperWithError.find('p')
    expect(p.text().trim()).toBe('ocurrio un error cargando los gifs')

  })

  it ('<GifGrid /> should have items when data is loaded', () => {

    const gifts = [{
      id: 1,
      title: 'first gift',
      url: 'https://someimage/image.gif'
    }]

    useFetchGifs.mockReturnValue({
      data: gifts,
      loading: false,
      error: false
    })
    const wrapperWithData = shallow(<GifGrid category={initialCategory}/>)


    expect(wrapperWithData).toMatchSnapshot()

    expect(wrapperWithData.find('p').exists()).toBeFalsy()
    expect(wrapperWithData.find('GifGridItem').length).toBe(gifts.length)

  })

})