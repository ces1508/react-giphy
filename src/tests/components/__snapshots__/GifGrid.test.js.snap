// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GifGrid /> <GifGrid /> should be rendered 1`] = `
<div
  className="animate__animated animate__fadeInDownBig animate__delay-2s"
>
  <h3>
    goku
  </h3>
  <p>
    Cargando ...
  </p>
  <div
    className="card-grids"
  />
</div>
`;

exports[`<GifGrid /> <GifGrid /> should have a p tag when useFetchGifs return an error 1`] = `
<div
  className="animate__animated animate__fadeInDownBig animate__delay-2s"
>
  <h3>
    goku
  </h3>
  <p>
    ocurrio un error cargando los gifs
  </p>
  <div
    className="card-grids"
  />
</div>
`;

exports[`<GifGrid /> <GifGrid /> should have items when data is loaded 1`] = `
<div
  className="animate__animated animate__fadeInDownBig animate__delay-2s"
>
  <h3>
    goku
  </h3>
  <div
    className="card-grids"
  >
    <GifGridItem
      id={1}
      key="1"
      title="first gift"
      url="https://someimage/image.gif"
    />
  </div>
</div>
`;
